pipeline/detection.py

```diff
@@ -706,6 +757,121 @@ def extract_sources_sep(self, image):
+
+        If the image has a "score" array, will use that.
+        If not, will apply the PSF of the image to make a score array.
+
+        Sources are detected as points in the score image that are
+        above the given threshold. If points are too close, they
+        will be merged into a single source.
+        The output SourceList will include the aperture and PSF photometry,
+        and the x,y positions will be given using the source centroid.
+
+        # TODO: should we do iterative PSF tapered centroiding?
+
+        Parameters
+        ----------
+        image: Image
+            The image to extract sources from. Must have a score, or a PSF,
+            that can be gotten from one of these places (in order):
+            - a PSF object loaded on the Image object itself.
+            - a zogy_psf attribute attached to the Image object.
+            - a PSF object of the new_image attribute of the Image object.
+            If the image also has a zogy_alpha attribute, it will be used
+            to extract the PSF photometry.
+
+        Returns
+        -------
+        sources: SourceList
+            The list of sources detected in the image.
+            This contains a table where each row represents
+            one source that was detected, along with all its properties.
+
+        """
+        score = image.score.copy()
+
+        if score is None:
+            raise NotImplementedError('Still need to add the matched-filter cross correlation! ')
+
+        psf_image = None
+        if image.psf is not None:
+            psf_image = image.psf.get_clip()
+        elif getattr(image, 'zogy_psf', None) is not None:
+            psf_image = image.zogy_psf
+
+        fwhm = image.fwhm_estimate
+        if fwhm is None:
+            fwhm = image.new_image.fwhm_estimate
+        if fwhm is None and image.psf is not None:
+            fwhm = image.psf.fwhm_pixels
+        if fwhm is None and image.new_image.psf is not None:
+            fwhm = image.new_image.psf.fwhm_pixels
+
+        if fwhm is None:
+            raise RuntimeError("Cannot find a FWHM estimate from the given image or its new_image attribute.")
+
+        # typical scale of PSF is X times the FWHM
+        fwhm_pixels = max(int(np.ceil(fwhm * self.pars.separation_fwhms / image.instrument_object.pixel_scale)), 1)
+        # remove flagged pixels
+        score[image.flags > 0] = np.nan
+
+        # remove the edges
+        border = fwhm_pixels * 2
+        score[:border, :] = np.nan
+        score[-border:, :] = np.nan
+        score[:, :border] = np.nan
+        score[:, -border:] = np.nan
+
+        # normalize the score based on sigma_clipping
+        # TODO: we should check if we still need this after b/g subtraction on the input images
+        mu, sigma = sigma_clipping(score)
+        score = (score - mu) / sigma
+        det_map = abs(score) > self.pars.threshold  # catch negative peaks too (can get rid of them later)
```

Comment:
I don't think you use it here, but it occurs to me that we might want two different thresholds.

The first one is the one where you find a single pixel that's above the threshold to preliminarily find objects.

The second one is the S/N inside the whole PSF (or inside an aperture, or whatever) for the objects that you keep.

(The extraction method we used in the SCP back in the late 90s did this two-step kind of thing.   I'm not sure in detail what sextractor does.)

Reply:
I'm not sure what the other threshold does. When match-filtering, each pixel already contains the aggregated S/N for the area of the PSF around that pixel. So looking at a single pixel is the right thing to do for source detection on a matched-filtered image.